### DISCUSSION TITLE
React to any remaining key (non bound ones)

### DISCUSSION THREAD

**Discussion by pawamoy** [Textualize/textual#3631] [Show and tell]
I sometimes push a modal that reacts to scroll keys (up, down, left, right, pageup, pagedown, home, end), and wanted to bind any remaining key to a "dismiss" action.

Turns out it's quite easy to do:

```python
class MyModal(ModalScreen):
    BINDINGS = [...]

    @property
    def bound_keys(self) -> Iterator[str]:
        for binding in self.BINDINGS:
            if isinstance(binding, Binding):
                yield from binding.key.split(",")
            else:
                yield from binding[0].split(",")

    def on_key(self, event: Key) -> None:
        if event.key not in self.bound_keys:
            self.dismiss()
```

Could be nice-to-haves in Textual itself:

- a `bindings` attribute that always returns actual `Binding` instances, and not tuples
- a `keys` attribute on `Binding` that does the comma split

This way `bound_keys` could be rewritten:

```python
for binding in self.bindings:
    yield from binding.keys
```

...and maybe this could become part of Textual too :slightly_smiling_face:

**Comment by davep** [comment]
Adjacent to #2269 too, I'd say. While not directly related, this feels like something to keep in mind if/when doing a full review of the binding system.

**Reply by pawamoy** [reply]
(it's a shame that there's no backlink functionality between issues and discussions...)

**Comment by davep** [comment]
As an aside, if you don't mind a sneaky dip into one internal on `App`, this should do the trick without needing to worry about splitting up the keys yourself, I think?

```python
from textual.app import App, ComposeResult
from textual.containers import VerticalScroll
from textual.events import Key
from textual.screen import ModalScreen
from textual.widgets import Static

class AnyKeyScreen(ModalScreen[None]):

    def compose(self) -> ComposeResult:
        with VerticalScroll():
            yield Static(
                "\n".join([
                    f"{n} on_key and no Binding make Jack a dull boy"
                    for n in range(1000)
                ])
            )

    def on_key(self, event: Key) -> None:
        if not any(bindings.keys.get(event.key) for _, bindings in self.app._binding_chain):
            self.dismiss()

class AnyKeyDismissApp(App[None]):

    def on_mount(self) -> None:
        self.push_screen(AnyKeyScreen())

if __name__ == "__main__":
    AnyKeyDismissApp().run()
```

Could also be that `self.app._modal_binding_chain` is more appropriate for your needs.

**Reply by pawamoy** [reply]
Thanks! I'll probably stay on beaten (trusted? known? blessed? what's the expression again?) paths and avoid using internals, but that's a good thing to know such attributes already exist :slightly_smiling_face: Maybe just a matter of exposing them then.

**Reply by davep** [reply]
Yeah, I was wondering if something akin to:

```python
def is_bound(self, key: Key | str) -> bool:
    key = key.key if isinstance(key, Key) else key
    return any(bindings.keys.get(key) for _, bindings in self._binding_chain)
```

might be a good idea for `App`.

**Reply by pawamoy** [reply]
```python
    def on_key(self, event: Key) -> None:
        if not self.app.is_bound(key):
            self.dismiss()
```

That'd be nice indeed.

**Reply by pawamoy** [reply]
Actually I went with your suggestion and `_modal_binding_chain`, it lets me write much less code. Let see what these internals become! Thanks again!
